File: 01.patterns/pattern-20.js
// example pattern
// *        *
// **      **
// ***    ***
// ****  ****
// **********
// ****  ****
// ***    ***
// **      **
// *        *

let n = 5
for(let i = 0; i < 2 * n - 1; i++){
    if(i < 5 ){
        let stars = ""
        for(let j = 0; j < i + 1; j++){
            stars += "*"
        }
    
        let spaces = ""
        for(let j = 0; j < n - i - 1; j++){
            spaces += " "
        }
        console.log(stars + spaces + spaces + stars)
    }else{
        let stars = ""
        for(let j = 0; j < 2 * n - i - 1; j++){
            stars += "*"
        }

        let spaces = ""
        for(let j = 0; j < i - n + 1; j++){
            spaces += " "
        }

        console.log(stars + spaces + spaces + stars)
    }
}

console.log("--- revision starts here ---")

// *        *
// **      **
// ***    ***
// ****  ****
// **********
// ****  ****
// ***    ***
// **      **
// *        *

let k = 5

for(let i = 0; i < 2 * k - 1; i++){
    let stars = ""
    for(let j = 0; i < k ? j < i + 1 : j < 2 * k - i - 1; j++){
        stars += "*"
    }

    let space = ""
    for(let j = 0; i < k ? j < k - i - 1 : j < i - k + 1; j++){
        space += " "
    }

    console.log(stars + space + space + stars)
}


// final revision
// *        *
// **      **
// ***    ***
// ****  ****
// **********
// ****  ****
// ***    ***
// **      **
// *        *

let rows = 5
for(let i = 0; i < 2 * rows - 1; i++){
    let stars = ''
    let spaces = ''
    if(i < rows){
        for(let j = 0; j < i + 1; j++){
            stars += '* '
        }

        for(let j = 0; j < rows - i - 1; j++){
            spaces += '  '
        }
    }else{
        for(let j = 0; j < 2 * rows - i - 1; j++){
            stars += '* '
        }

        for(let j = 0; j < i - rows + 1; j++){
            spaces += '  '
        }
    }

    console.log(stars + spaces + spaces + stars)
}

// practise for a job application
console.log("-----for job application-----")

// *        *
// **      **
// ***    ***
// ****  ****
// ********** 4
// ****  **** 5 1
// ***    *** 6 2
// **      ** 7 3
// *        * 8 4

let p = 5

for(let i = 0; i < 2 * p - 1; i++){
    
    let space = ''
    let row = ''

    if(i < p){
        for(let j = 0; j < i + 1; j++){
            row += '*'
        }

        for(let j = 0; j < p - i - 1; j++){
            space += ' '
        }
    }else{
        for(let j = 0; j < 2 * p - i - 1; j++){
            row += '*'
        }

        for(let j = 0; j < i - p + 1; j++){
            space += ' '
        }
    }

    console.log(row + space + space + row)
}

File: 01.patterns/pattern-20.go
// ------------------ revision with js and go -------------------------
package main

import (
	"fmt"
	"strings"
)

// *        *
// **      **
// ***    ***
// ****  ****
// ********** 4
// ****  **** 5 1
// ***    *** 6 2
// **      ** 7 3
// *        * 8 4

func main() {
	p := 5

	// // for i := 0; i < p; i++ {

	// // }

	// for i := 0; i < p; i++ {
	// 	space := strings.Repeat("  ", p-i-1)
	// 	stars := strings.Repeat("* ", 2*i+1)
	// 	fmt.Println(space + stars)
	// }

	// for i := 0; i < 2*p-1; i++ {
	// 	var stars string
	// 	if i < p {
	// 		stars = strings.Repeat("* ", i+1)
	// 	} else {
	// 		stars = strings.Repeat("* ", 2*p-i-1)
	// 	}
	// 	fmt.Println(stars)
	// }

	// *        *
	// **      **
	// ***    ***
	// ****  ****
	// ********** 4
	// ****  **** 5 1
	// ***    *** 6 2
	// **      ** 7 3
	// *        * 8 4

	fmt.Println("---------")
	for i := 0; i < 2*p-1; i++ {
		var b strings.Builder
		var b2 strings.Builder

		if i < p {
			starCount := i + 1
			spaceCount := p - starCount

			for j := 0; j < starCount; j++ {
				b.WriteByte('*')
			}
			for j := 0; j < spaceCount; j++ {
				b2.WriteByte(' ')
			}
		} else {
			spaceCount := i - p + 1
			starCount := p - spaceCount

			for j := 0; j < starCount; j++ {
				b.WriteByte('*')
			}
			for j := 0; j < spaceCount; j++ {
				b2.WriteByte(' ')
			}
		}

		fmt.Println(b.String() + b2.String() + b2.String() + b.String())
	}

}
